IdentityCore/src/webview/embeddedWebview/MSIDOAuth2EmbeddedWebviewController.m

```diff
@@ -26,21 +26,375 @@
+        NSURL *url = navigationAction.request.URL;
+        [self webAuthDidCompleteWithURL:url];
+        
+        decisionHandler(WKNavigationActionPolicyCancel);
+        return;
+    }
+    
+    if ([requestUrlString isEqualToString:@"about:blank"])
+    {
+        decisionHandler(WKNavigationActionPolicyAllow);
+        return;
+    }
+    
+    if ([[[requestUrl scheme] lowercaseString] isEqualToString:@"browser"])
+    {
+        self.complete = YES;
+        requestUrlString = [requestUrlString stringByReplacingOccurrencesOfString:@"browser://" withString:@"https://"];
+        
+#if TARGET_OS_IPHONE
+        if (![MSIDAppExtensionUtil isExecutingInAppExtension])
+        {
+            [self cancel];
+            [MSIDAppExtensionUtil sharedApplicationOpenURL:[[NSURL alloc] initWithString:requestUrlString]];
+        }
+        else
+        {
+            MSID_LOG_INFO(self.context, @"unable to redirect to browser from extension");
+        }
+#else
+        [[NSWorkspace sharedWorkspace] openURL:[NSURL URLWithString:requestUrlString]];
+#endif
+        
+        decisionHandler(WKNavigationActionPolicyCancel);
+        return;
+    }
+    
+#if TARGET_OS_IPHONE
+    // check for pkeyauth challenge.
+    if ([requestUrlString hasPrefix:[kMSIDPKeyAuthUrn lowercaseString]])
+    {
+        decisionHandler(WKNavigationActionPolicyCancel);
+        [MSIDPKeyAuthHandler handleChallenge:requestUrl.absoluteString
+                                     context:self.context
+                           completionHandler:^(NSURLRequest *challengeResponse, NSError *error) {
+                               if (!challengeResponse)
+                               {
+                                   [self endWebAuthenticationWithError:error orURL:nil];
+                                   return;
+                               }
+                               [self loadRequest:challengeResponse];
+                           }];
+        return;
+    }
+#endif
+    
+    // redirecting to non-https url is not allowed
+    if (![[[requestUrl scheme] lowercaseString] isEqualToString:@"https"])
+    {
+        MSID_LOG_INFO(self.context, @"Server is redirecting to a non-https url");
+        self.complete = YES;
+        
+        NSError *error = MSIDCreateError(MSIDErrorDomain, MSIDServerNonHttpsRedirect, @"The server has redirected to a non-https url.", nil, nil, nil, self.context.correlationId, nil);
+        [self endWebAuthenticationWithError:error orURL:nil];
+        
+        decisionHandler(WKNavigationActionPolicyCancel);
+        return;
+    }
+    
+    decisionHandler(WKNavigationActionPolicyAllow);
+}
+
+- (void)webView:(WKWebView *)webView didStartProvisionalNavigation:(null_unspecified WKNavigation *)navigation
+{
+    if (!self.loading)
+    {
+        self.loading = YES;
+        if (_spinnerTimer)
+        {
+            [_spinnerTimer invalidate];
+        }
+        _spinnerTimer = [NSTimer scheduledTimerWithTimeInterval:2.0
```

Comment:
Just curious, why this is necessary?

Reply:
Showing spinner only if loading takes more than 2 seconds. It sounds like a better experience, what do you think?